src/main/java/com/ververica/flink/table/gateway/deployment/ClusterDescriptorAdapter.java

```diff
@@ -0,0 +1,155 @@
+            } catch (Throwable t) {
+                // the job might has finished earlier
+            }
+            return null;
+        });
+    }
+
+    /**
+     * The reason of using ClusterClient instead of JobClient to retrieve a cluster is
+     * the JobClient can't know whether the job is finished on yarn-per-job mode.
+     *
+     * <p>If a job is finished, JobClient always get java.util.concurrent.TimeoutException
+     * when getting job status and canceling a job after job is finished.
+     * This method will throw org.apache.hadoop.yarn.exceptions.ApplicationNotFoundException
+     * when creating a ClusterClient if the job is finished. This is more user-friendly.
+     */
+    protected <R> R bridgeClientRequest(
+            ExecutionContext<ClusterID> executionContext,
+            JobID jobId,
+            String sessionId,
+            Function<ClusterClient<?>, R> function) {
+        if (this.clusterID == null) {
+            LOG.error("Session: {}. Cluster information don't exist.", sessionId);
+            throw new IllegalStateException("Cluster information don't exist.");
+        }
+        // stop Flink job
+        try (final ClusterDescriptor<ClusterID> clusterDescriptor = executionContext.createClusterDescriptor()) {
+            try (ClusterClient<ClusterID> clusterClient =
+                         clusterDescriptor.retrieve(this.clusterID).getClusterClient()) {
+                // retrieve existing cluster
+                return function.apply(clusterClient);
+            } catch (Exception e) {
+                LOG.error(
+                        String.format("Session: %s, job: %s. Could not retrieve or create a cluster.", sessionId, jobId),
+                        e);
+                throw new SqlExecutionException("Could not retrieve or create a cluster.", e);
+            }
+        } catch (SqlExecutionException e) {
+            throw e;
+        } catch (Exception e) {
+            LOG.error(
+                    String.format("Session: %s, job: %s. Could not locate a cluster.", sessionId, jobId), e);
+            throw new SqlExecutionException("Could not locate a cluster.", e);
+        }
+    }
+
+    /**
+     * Set the flink job id.
+     */
+    public void setJobId(JobID jobId) {
+        this.jobId = jobId;
+    }
+
+    /**
+     * Checks whether this job state is <i>globally terminal</i>.
+     * A globally terminal job is complete and cannot fail any more
+     * and will not be restarted or recovered by another standby master node.
+     *
+     * When a globally terminal state has been reached,
+     * all recovery data for the job is dropped from the high-availability services.
+     *
+     * @return True, if this job status is globally terminal, false otherwise.
+     */
+    public abstract boolean isGloballyTerminalState();
+
+    /**
+     * Set ClusterID.
+     */
+    public abstract void setClusterID(Configuration configuration);
```

Comment:
Also move this `configuration` into constructor. `Configuration` should also be reused for the same job under a high-availability scenario. See the top-most comment.